### DISCUSSION TITLE
Testing a `wgpu` application

### DISCUSSION THREAD

**Discussion by zmitchell** [gfx-rs/wgpu#4182] [Q&A] [answered]
I'm basically brand new to any kind of graphics, so apologies if there are standard answers for this kind of thing.

I'm wondering what a typical test setup might look like for a `wgpu` application. For context I'm working on a fork of `nannou`, a creative coding framework (think small animations, 2D art, etc).

For example, what kinds of tests even make sense in a graphics context? If I want to test a command that says "draw a circle", would the test render the circle in a solid color, read the texture back into CPU memory, then check a pixel inside the circle to see if it's the right color? Is that the general strategy? If I want to take the "middleware" approach (outlined in this repo) for a part of my application, is there a way to determine that I haven't accidentally called `submit` somewhere?

Anyway, this isn't a `wgpu`-specific question, I just figured that the maintainers would have good insight into this.

**Comment by CatCode79** [comment]
I personally can't help you, but I was very impressed with the [all-is-cubes renderer](https://github.com/kpreid/all-is-cubes/tree/main) test system created by @kpreid, I think taking a look at it might help you.

Rust Gamedev Meetup video:
https://youtu.be/s9kf9HVUKYE?list=PLYiOdhpKxxXI9l8V15FciLcsPzkz3Gt4I&t=4620
Related crate:
https://github.com/kpreid/all-is-cubes/tree/main/test-renderers
Another useful crate:
https://github.com/kpreid/rendiff/

**Comment by cwfitzgerald** [comment, marked as answer]
> If I want to test a command that says "draw a circle", would the test render the circle in a solid color, read the texture back into CPU memory, then check a pixel inside the circle to see if it's the right color? Is that the general strategy? 

That's quite a common solution. We do this within wgpu itself and I do it in my library [rend3](https://github.com/BVE-Reborn/[rend3](https://github.com/BVE-Reborn/rend3/)). The devil is in the details though. wgpu and rend3 use bindings to an nvidia library called [FLIP](https://github.com/gfx-rs/nv-flip-rs) which does human-style image comparisons. I personally am quite a fan of it for how well it deals with noisy inputs or inputs with very low variance.

**Comment by zmitchell** [comment]
Awesome, thanks both of you

**Comment by zmitchell** [comment]
I thought that `flip` package was cool so I have a PR up to include `flip` in `nixpkgs`: https://github.com/NixOS/nixpkgs/pull/262183

It's only the non-CUDA binary (none of the Python stuff), but it will let anyone consume the package from Nix.
